packages/inferno-animation/src/AnimatedComponent.ts

```diff
@@ -0,0 +1,96 @@
+import { Component } from 'inferno'
+import {
+  addClassName,
+  removeClassName,
+  registerTransitionListener,
+  forceReflow,
+  clearDimensions,
+  getDimensions,
+  setDimensions,
+} from './utils'
+
+type Props = {
+  animation: string|object
+}
+
+export default class AnimatedComponent extends Component<Props> {
+  didAppear (dom) {
+    let animCls: any; // TODO: This should be typed properly
+    if (typeof this.props.animation === 'object') {
+      animCls = this.props.animation;
+    }
+    else {
+      const animationName = this.props.animation || 'inferno-animation';
+      animCls = {
+        'start': animationName + '-enter',
+        'active': animationName + '-enter-active',
+        'end': animationName + '-enter-end'
+      }
+    }
+  
+    // 1. Get height and set start of animation
+    const { width, height } = getDimensions(dom)
+    addClassName(dom, animCls.start)
+    forceReflow()
+  
+    // 2. Activate transition
+    addClassName(dom, animCls.active)
+  
+    // 3. Set an animation listener, code at end
+    // Needs to be done after activating so timeout is calculated correctly
+    registerTransitionListener([dom, dom.children[0]], function () {
+      // *** Cleanup ***
+      // 5. Remove the element
+      clearDimensions(dom)
+      removeClassName(dom, animCls.active)
+      removeClassName(dom, animCls.end)
+      
+      // 6. Call callback to allow stuff to happen
+      // callback(node)
+    }, false)
+  
+    // 4. Activate target state
+    setTimeout(() => {
+      setDimensions(dom, width, height)
+      removeClassName(dom, animCls.start)
+      addClassName(dom, animCls.end)
+    }, 5)
+  }
+
+  willDisappear (dom, callback) {
+    let animCls;
+    if (typeof this.props.animation === 'object') {
+      animCls = this.props.animation;
+    }
+    else {
+      const animationName = this.props.animation || 'inferno-animation';
+      animCls = {
+        'start': animationName + '-leave',
+        'active': animationName + '-leave-active',
+        'end': animationName + '-leave-end'
+      }
+    }
+  
+    // 1. Get dimensions and set animation start state
+    const { width, height } = getDimensions(dom)
+    setDimensions(dom, width, height)
+    addClassName(dom, animCls.start)
+    
+    // 2. Activate transitions
+    addClassName(dom, animCls.active);
+  
+    // 3. Set an animation listener, code at end
+    // Needs to be done after activating so timeout is calculated correctly
+    registerTransitionListener(dom, function () {
+      // *** Cleanup not needed since node is removed ***
+      callback()
+    }, false)
+  
+    // 4. Activate target state
+    setTimeout(() => {
```

Comment:
what is the reason for timeout and why 5ms?

Reply:
@Havunen the setTimeout is needed otherwise the browser engine optimises the following sequence of operations:

```
// 2. Activate transition
addClassName(dom, animCls.active)
// 4. Activate target state
setDimensions(dom, width, height)
removeClassName(dom, animCls.start)
addClassName(dom, animCls.end)
```

and immediately goes to the end state without a transition. Setting it too short also triggers optimisation. I could switch to requestAnimationFrame, I believe the choice of setTimeout was to support IE9 without polyfill a couple of years back.

I will switch to requestAnimationFrame unless you have another opinion.

Reply:
Yeah, IE9 and IE10 are not officially supported by inferno js and I personally hope all users stop using them.

Reply:
FIXED in aeebd7240